### DISCUSSION TITLE
Use different list/queue for different text that ends up in the message bar

### DISCUSSION THREAD

**Discussion by montaropdf** [stumpwm/stumpwm#1113] [General]
Hello,

I am a user of `which-key-mode` and multiple contribution modules. `which-key-mode` and some of those modules display text in the message bar and its queue/list. However, error messages also ends up into the message bar and its queue/list.

When using the function `lastmsg` to try going back in the list of messages, all those texts are clobbering the list of messages. Moreover, with `which-key-mode` activate, it is impossible to go back, as a new message is created with the list of available bindings, before I can press the `m` key.

So, error messages and StumpWM "core" should use a different queue/list and a different set of functions to read it.

Another good feature would be to provide a window displaying the content of one of those lists.

**Comment by szos** [comment]
I'm unsure what the wontfix label means in the context of stumpwm, but given that this issue was also tagged with enhancement I'm assuming it just means that the maintainers dont intend to fix this, and not that PR's addressing this wont be accepted. 

@montaropdf, you can prevent messages from being added by setting `*record-last-msg-override*` to `t`. For example, you can prevent `which-key-mode` messages from being added to the screens list of messages like so: 

```
(defun which-key-mode-key-press-hook (key key-seq cmd)
  "*key-press-hook* for which-key-mode"
  (declare (ignore key cmd))
  (when (not (eq *top-map* *resize-map*))
    (let* ((*record-last-msg-override* t)
           (oriented-key-seq (reverse key-seq))
           (maps (get-kmaps-at-key-seq (dereference-kmaps (top-maps)) oriented-key-seq)))
      (when-let ((only-maps (remove-if-not 'kmap-p maps)))
        (apply 'display-bindings-for-keymaps oriented-key-seq only-maps)))))
```

**Comment by montaropdf** [comment]
@szos , thank you for the code snippet.

Recently, I think again about this, and I am wondering if bringing the concept of buffers like in emacs or nyxt, wouldn't help decouple more easily data to be retained during the course of a stumpwm session, from "notifications", generated by `which-key` or `app-menu`.

With such design, messages to be retained, would be written to the `*Messages` buffer (or any other buffer name) and a function to display the message, could be attached to a hook triggered whenever a new piece of text is append to the `*Message*` buffer, for example.

Buffers are, maybe, easier to inspect, from a REPL, than the current situation, because buffers would be a generic class, with some specialised subclasses, when needed. Currently, I don't know how to inspect the content of the current "queue" of messages displayed.

Buffers shouldn't enforce a particular data format, but should provide as much information as possible about the format of the stored data. This could ease inspection, via a REPL, or any other tool dedicated to such task. It could also allow various pieces of code, to share the same buffer, when it make sense, of course.

Issue #779, #838, and associated issues, should be part of further discussion on this topic, as they could be impacted.

Implementation wise, it is not clear to me, how buffers would fit into the current system. It is not clear how they would affect configuration, either. As such implementation would be, surely, a big change in the design of stumpwm, it doesn't seem excessive, to me, to impose a change in one's configuration files.

**Comment by szos** [comment]
This seems like a big thing to implement to solve the issue of messages being clobbered. I think that specific issue could be addressed by reworking how the message list is walked, or adding new functions to walk the list in a new way, or just preventing certain messages from being collected into the list. I dont think having different lists is necessary to address the issue.

Personally, I think the overhead of implementing buffers isnt worthwhile at this point in time. The overhead of changing how messages are handled is more expensive (imo) than changing how the message list is walked, and the added benefit seems (please correct me if im wrong) to be categorization of messages, which isnt necessarily best solved by dividing them up in different lists. 

Also, imposing a change in peoples configs is (again imo) excessive. Backwards compatibility is important, and when someone upgrades StumpWM and suddenly their config isnt working, or behavior has changed in non-trivial ways, thats a big deal for them.

**Comment by montaropdf** [comment]
I should, maybe, open a dedicated issue for the buffer implementation, I see it as a generic way to store and handle data, like messages, or any data a module would like to handle, like data from a database or a picture, to be displayed in a dedicated window, etc. I wanted to bring the discussion about buffers, here, because, messages handling could be affected.

As I don't know the internals of StumpWM, it is difficult, for me, to propose something specific and limited in scope. I only see this problem from the lense of a regular user. 

I understand the reluctance to breaking backward compatibility and I agree, implementing buffers would be a big design change, yet, this is also a question of warning users properly, before hand. I suppose this would means, at the very least changing the major version of StumpWM.

So let's continue with your proposal to change the way the message list is walked, I want to know more about what you think could be done to improve the situation.

**Comment by szos** [comment]
Regarding changing how the message list is walked, I've taken a peek at `lastmsg` and implemented a simple walker that keeps its place in the message list, but requires a manual reset. This is pretty much what lastmsg does already, just broken into two commands. I really do think that the better solution is to bind `*record-last-msg-override*` to `t` where appropriate. heres the 5 minute code i threw together: 

```
(let ((lastmsg-cons nil)
      (lasthlt-cons nil))
  (defcommand saving-lastmsg () ()
    (let ((*record-last-msg-override* t)
          (screen (current-screen)))
      (if (car lastmsg-cons)
          (progn
            (apply 'echo-string-list screen (car lastmsg-cons) (car lasthlt-cons))
            (pop lastmsg-cons)
            (pop lasthlt-cons))
          (message "No more messages."))))
  (defcommand saving-lastmsg-reset () ()
    (let ((screen (current-screen)))
      (setf lastmsg-cons (screen-last-msg screen)
            lasthlt-cons (screen-last-msg-highlights screen)))))
```

If youre interested in learning the internals of StumpWM, imo the best way to do so is with by using emacs+sly(or slime) and using `M-.`, `M-,`, and `M-x sly-who-calls` to see how everything is strung together. 

As far as StumpWM versioning goes, It seems like new versions are released about once a year, in the form XX.yy. I know many folks clone directly from master. I think as far as this idea of buffers goes, it would be best implemented with CLIM application frames (or similar, im just familiar with CLIM so thats my go-to) on a per module basis. This would allow StumpWM to keep its behavior, and allow each module to specify its own windows behavior. But like you said, thats outside the scope of this issue.

**Comment by tychoish** [comment]
Is there any particular reason why stumpwm needs to be responsible for this data at all? I agree with @szos that buffers is "a big thing" (though I do like the idea,) 

I do all of my personal notifications using dunst, (and use dmenu for most shell/etc sthings.) and while I use the standard (ish) modeline, and sometimes think about making it more spiffy, I sort think that it'd be better to write a mode-line/tray that could/would run in a separate process. Which would be interesting because maybe non-stumpwm people would use it, the UI for a tray is pretty distinct from the problem of "being an x11 window manager," and isolating tray-type things in their own process could help stability (mostly because the most volatile/rough user code that runs in stump is probably mode-line related, so making it possible for the mode line to crash hard without crashing the WM seems good.

**Comment by szos** [comment]
I dont think StumpWM needs to be responsible for this data, rather the specific module should be responsible for itself. I cant think of any modules offhand that would necessitate a buffer/application frame/GUI, but I suppose montaropdf has some that they think could use it. 

Regarding the mode line, There is some work being done on a CLIM mode line, but as youve noted, It has some issues with being run in a separate process within the same image.
